storage/remote/metadata_watcher.go

```diff
@@ -0,0 +1,158 @@
+)
+
+// MetadataAppender is an interface used by the Metadata Watcher to send metadata read from the scrape manager on to somewhere else.
+type MetadataAppender interface {
+	AppendMetadata(context.Context, []scrape.MetricMetadata)
+}
+
+// Watchable is an interface used to represent a scrape manager from which we fetch metadata.
+type Watchable interface {
+	TargetsActive() map[string][]*scrape.Target
+}
+
+// MetadataWatcher watches the Scrape Manager for a given WriteMetadataTo.
+type MetadataWatcher struct {
+	name   string
+	logger log.Logger
+
+	managerGetter scrape.ManagerGetter
+	manager       Watchable
+	writer        MetadataAppender
+
+	interval model.Duration
+	deadline time.Duration
+
+	quit chan struct{}
+	done chan struct{}
+
+	hardShutdown context.CancelFunc
+}
+
+// NewMetadataWatcher builds a new MetadataWatcher.
+func NewMetadataWatcher(l log.Logger, mg scrape.ManagerGetter, name string, w MetadataAppender, interval model.Duration, deadline time.Duration) *MetadataWatcher {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	if mg == nil {
+		mg = &scrape.ReadyManager{}
+	}
+
+	return &MetadataWatcher{
+		name:   name,
+		logger: l,
+
+		managerGetter: mg,
+		writer:        w,
+
+		interval: interval,
+		deadline: deadline,
+
+		quit: make(chan struct{}),
+		done: make(chan struct{}),
+	}
+}
+
+// Start the MetadataWatcher.
+func (mw *MetadataWatcher) Start() {
+	level.Info(mw.logger).Log("msg", "Starting scraped metadata watcher")
+	var hardShutdownCtx context.Context
+	hardShutdownCtx, mw.hardShutdown = context.WithCancel(context.Background())
+
+	go mw.loop(hardShutdownCtx)
+}
+
+// Stop the MetadataWatcher.
+func (mw *MetadataWatcher) Stop() {
+	level.Info(mw.logger).Log("msg", "Stoping metadata watcher...")
+	defer level.Info(mw.logger).Log("msg", "Scraped metadata watcher stopped")
+
+	close(mw.quit)
+	select {
+	case <-mw.done:
+		return
+	case <-time.After(mw.deadline):
+		level.Error(mw.logger).Log("msg", "Failed to flush metadata")
+	}
+
+	mw.hardShutdown()
+	<-mw.done
+}
+
+func (mw *MetadataWatcher) loop(ctx context.Context) {
+	ticker := time.NewTicker(time.Duration(mw.interval))
+	defer ticker.Stop()
+	defer close(mw.done)
+
+	for {
+		select {
+		case <-mw.quit:
+			return
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			mw.collect(ctx)
+		}
+	}
+}
+
+func (mw *MetadataWatcher) collect(ctx context.Context) {
+	if !mw.readyForCollection() {
+		return
+	}
+
+	metadata := []scrape.MetricMetadata{}
+	for _, targets := range mw.manager.TargetsActive() {
+		for _, tt := range targets {
+			metadata = append(metadata, tt.MetadataList()...)
+		}
+	}
+
+	// Blocks until the metadata is sent to the remote write endpoint or closed.
+	mw.writer.AppendMetadata(ctx, metadata)
+}
+
+func (mw *MetadataWatcher) readyForCollection() bool {
+	if mw.manager != nil {
+		return true
+	}
+
+	if err := mw.setManager(); err != nil {
+		return false
+	}
+
+	return true
+}
+
+func (mw *MetadataWatcher) setManager() error {
```

Comment:
do we need this function? Looks shallow - used only once. Can we inline instead?